design/2025-10-persistent-replica-runtime.md

```diff
@@ -0,0 +1,258 @@
+
+These are ensured to exist by the replica description.
+
+### Replica actor
+
+A replica actor is responsible for handling input data, calling revision
+reducers, and persisting new revisions and deltas.
+
+A revision actor is produced by a replica description and is instantiated with
+partition data.
+
+### Replica description
+
+A replica description describes how to build replicas from the partition data,
+and also how to build revision actors.
+
+### Rebuild delta
+
+A rebuild delta is a specific kind of delta that is the direct result of a
+change in implementation of a replica. For instance it may be important to issue
+a rebuild delta when bugs are fixed in replica descriptions that meant replica's
+were producing faulty data.
+
+### Partition
+
+replicas are partitioned by a key. This key is usually something like a Matrix
+room ID, or the user identifier of a Draupnir instance.
+
+### Actor ID
+
+These prevent multiple live in-memory actors providing for the same replica. The
+scheduler makes sure that there is only ever one live actor per replica
+partition. This does require the scheduler to rebuild all actors if the
+scheduler crashes, but i'm fine with that.
+
+### Revision ID
+
+A revision identifier is a specific ULID that is associated with a revision.
+
+### Intent replica (effect)
+
+A intent replica is a special kind of replica that produces an intention to see
+some outcome enacted. These intents are then carried out by outcome issuers
+(effect handler).
+
+Enabled protections should be intent replicas. These protections are partitioned
+by the capability provider set they are using and the draupnir they are
+associated with. They are not associated with the in-memory instance but the
+activation of the protection.
+
+### Outcome replica (effect handler)
+
+An outcome replica is a special kind of replica that records whether actions
+have been issued for intents. The semantic meaning of the outcome is not
+relevant, e.g. failure to ban a user is still an outcome that needs to be
+recorded and is the outcome of the intent.
+
+outcome replicas for protections are partitioned in the same way as the intent
+replicas, it's just the capability providers are responsible for producing the
+actions.
+
+## Semantics
+
+### Actor migration
+
+Actors need to be transportable across node.js worker threads in order to
+balance CPU.
+
+### Preemption
+
+replica actor client code will be interruptive in order to gracefully handle
+shutdown and migration of actors.
+
+This will be achieved by abusing I/O operations to signal interrupts. CPU bound
+actor code will still have to cooperatively check for an interrupt signal.
+
+### Garbage collection
+
+Replica dependents will formally be tracked. Transitive dependencies will also
+need to be considered for rollbacks to be successful. replicas will only be able
+to discard an old revision when all transitive dependents have successfully
+migrated to a superseding revision.
```

Comment:
Hm, we need to also think about how some replicas interact with this. Like some protections and capabilities may no longer be relevant to keep around? Especially protections that have been instantiated with the intent to observe their effects and nothing more... it needs some thought.